next/utils/newBooksService.ts

```diff
@@ -0,0 +1,164 @@
+      continue
+    }
+    try {
+      // Image is already checked.
+      if (book.coverUrl in bookCoverMap) {
+        // eslint-disable-next-line no-continue
+        continue
+      }
+      // `url-file-size` gets only filesize information from the server, there's no need to download
+      // the whole image.
+      // eslint-disable-next-line no-await-in-loop
+      const coverFileSize = await ufs(book.coverUrl)
+
+      // Checking image size in bytes is a reliable way how to detect a generic cover.
+      bookCoverMap[book.coverUrl] = coverFileSize !== 9149
+
+      // eslint-disable-next-line no-empty
+    } catch (error) {}
+  }
+}
+
+const fetchBooks = async () => {
+  // In localhost environments mock books are returned not to make unnecessary request as the service
+  // is restarted with each compilation.
+  if (process.env.NEXT_PUBLIC_IS_LOCALHOST === 'true') {
+    fetchedBooks = mockNewBooks
+    return
+  }
+
+  try {
+    const response = await fetch(bookNewsUrl)
+    const text = await response.text()
+    const opac: OpacResponse = XMLConvertor.xml2js(text, { compact: true }) as OpacResponse
+
+    fetchedBooks = opac.xml.book.map((book) => {
+      /* eslint-disable no-underscore-dangle */
+      const coverUrl = book?.coverURL?._text
+      return {
+        title: book?.title?._text,
+        url: book?.recURL?._text,
+        ...(coverUrl === ' --- ' ? {} : { coverUrl }),
+      }
+      /* eslint-enable no-underscore-dangle */
+    })
+    // Check is called asynchronously without waiting for a result.
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    checkCoverImages()
+    // eslint-disable-next-line no-empty
+  } catch (error) {}
+}
+
+let refreshIntervalTimer: NodeJS.Timer | null = null
+
+export interface NewBooksResponse {
+  books: Book[]
+  total: number
+}
+
+export const getNewBooks = async ({
+  page,
+  pageSize,
+}: {
+  page?: number
+  pageSize?: number
+}): Promise<NewBooksResponse> => {
+  if (!refreshIntervalTimer) {
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    refreshIntervalTimer = setInterval(() => fetchBooks(), 3_600_000)
+  }
+
+  if (!fetchedBooks) {
+    // This might trigger multiple requests when `fetchedBooks` are not initialized yet.
+    await fetchBooks()
+  }
+
+  const books =
+    fetchedBooks?.map((book) => {
+      if (!book.coverUrl) {
+        return book
+      }
+
+      const imageChecked = book.coverUrl in bookCoverMap
+      // As the check is run asynchronously without waiting for a result, the cover might not be checked,
+      // in that case we return the cover even if it's generic.
+      const coverUrl = imageChecked
+        ? bookCoverMap[book.coverUrl]
+          ? book.coverUrl
+          : undefined
+        : book.coverUrl
+      const result = {
+        ...book,
+        coverUrl,
+      }
+      if (!coverUrl) {
+        // `undefined` is not serializable by Next getStaticProps
```

Comment:
Maybe we can return `null` instead of `undefined` and then it is searializable and you don't have to take special care? :D